rfc/2020/conformance-validation/mayankshah1607/rfc.md

```diff
@@ -0,0 +1,477 @@
+
+**1. Automatic Proxy Injection**
+
+Proxy injection process works by adding a `linkerd.io/inject: enabled` annotation to pod template / namespace. This triggers an admission webhook that injects the `linkerd-proxy` container to the targeted deployments. This process shall be tested in 7 steps:
+- Check if `inject` command (with and without `--manual` flag) outputs the desired configuration with the appended annotation. We shall provide an injected YAML file of the deployments which can be compared against the YAML blob output of inject.
+
+Alternatively, as users may want to test for conformance on their own sample applications (check "Future possibilities" section), we may have to make use of [kubernetes/client-go](https://github.com/kubernetes/client-go) to only check for the existence of the `linkerd.io/inject: enabled` annotation under the pod metadata.     
+- Wait for the pods to come to a _Running_ state for a certain duration, after which the test shall fail.
+- Check if the pods actually have the `linkerd-proxy` and the `linkerd-init` container
+- Check if the `linkerd-proxy` container is reachable by sending a `GET` request to liveness/readiness probe - to ensure service discovery is functional.
+- Check for any `SkipInject` errors in events
+- Check if uninject command updates the annotation to `linkerd.io/inject: disabled`
+- Wait for the pods to come to a _Running_ state for a certain duration, after which the test shall fail.
+- Ensure that the pods no longer have the `linkerd-init` and `linkerd-proxy` containers.
+
+**2. Essential linkerd commands**
+
+**linkerd tap cmd**
+
+This test will work by issuing the `linkerd tap` cmd on various resources of the sample application. The output returned shall be checked for any errors. For e.g - GKE may require extra RBAC configurations and permissions. Tap shall be tested by issuing a tap command using the `-o json` flag and performing checks on the obtained JSON :
+- Check for gRPC status codes
+- Check authority
+- Check for tls settings
+- Check httpStatus
+
+**linkerd stat**
+
+- Check if prometheus pod(s) is/are available
+- Check if controller pods are up and running
+- Check if the application pods are up and running
+- Issue stat cmd and validate output by checking for success rate and various latencies 
+> Note: This test may have to be put on hold due to https://github.com/linkerd/linkerd2/pull/3693
+
+**linkerd edges**
+
+- Check if application pods are up and running
+- Check if application deployment has desired replicas
+- Issue `linkerd edge` cmd using `-o json` flag, on various resources
+- Validate the `no_tls_reason` field of JSON output to ensure that the edges are mTLS'ed.
+
+**linkerd routes**
+- Check if application pods are up and running
+- Issue `linkerd routes` cmd
+- Validate the returned output by counting instances of desired route substrings using `strings.Count`
+  
+**3. Ingress**
```

Comment:
This is a good test to have. Two things to keep in mind:

1. If this test fails, the test needs to be able to report useful information back to the users, to aid with debugging. Or maybe we can just keep the cluster around so users can e.g. retrieve the ingress controller logs (in debug mode)
1. Local K8s clusters won't return an external IP